File: lintcode/python/0202_segment_tree_query.py
#!/usr/bin/python -t

# segment tree query

# beats 100%

"""
Definition of SegmentTreeNode:
class SegmentTreeNode:
    def __init__(self, start, end, max):
        self.start, self.end, self.max = start, end, max
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of segment tree.
    @param start: start value.
    @param end: end value.
    @return: The maximum number in the interval [start, end]
    """
    def query(self, root, start, end):
        # write your code here
        if root.start == start and root.end == end:
            return root.max
            
        mid = (root.start+root.end)/2
        left, right = -sys.maxint, -sys.maxint
        #print root.start, root.end, mid
        #print start, end
        # left side
        if start <= mid:
            if mid < end:
                left = self.query(root.left, start, mid)
            else:
                left = self.query(root.left, start, end)
                
        # right side
        if mid < end:
            #print "right side ", start, end
            if start <= mid:
                right = self.query(root.right, mid+1, end)
            else:
                right = self.query(root.right, start, end)
                
        # or not within root start and end area.
        
        return max(left, right)
                

"""
Definition of SegmentTreeNode:
class SegmentTreeNode:
    def __init__(self, start, end, max):
        self.start, self.end, self.max = start, end, max
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of segment tree.
    @param start: start value.
    @param end: end value.
    @return: The maximum number in the interval [start, end]
    """
    def query(self, root, start, end):
        # write your code here
        if root == None or start > end:
            return 0
            
        if start <= root.start and root.end <= end:
            return root.max
            
        return max(self.query(root.left, start, end),self.query(root.right, start, end))

File: lintcode/golang/0202_segment_tree_query.go
/**
 * Definition of SegmentTreeNode:
 * type SegmentTreeNode struct {
 *     Start, End, Max int
 *     Left, Right     *SegmentTreeNode
 * }
 */

/**
 * @param root: The root of segment tree.
 * @param start: start value.
 * @param end: end value.
 * @return: The maximum number in the interval [start, end]
 */

const MaxInt = int(^uint(0) >> 1)
const MinInt = -MaxInt-1

func query (root *SegmentTreeNode, start int, end int) int {
    // write your code here
    if root == nil {
        return 0
    }
    
    if start == root.Start && end == root.End {
        return root.Max
    }
    
    mid := (root.Start+root.End)/2
    left := MinInt
    right := MinInt
    
    if start <= mid {
        if mid < end {
            left = query(root.Left, start, mid)
        } else {
            left = query(root.Left, start, end)
        }
    }
    
    if mid < end {
        if start <= mid {
            right = query(root.Right, mid+1, end)
        } else {
            right = query(root.Right, start, end)
        }
    }
    
    return max(left, right)
}

func max (a, b int) int {
    if a > b {
        return a
    }
    return b
}

/**
 * Definition of SegmentTreeNode:
 * type SegmentTreeNode struct {
 *     Start, End, Max int
 *     Left, Right     *SegmentTreeNode
 * }
 */

/**
 * @param root: The root of segment tree.
 * @param start: start value.
 * @param end: end value.
 * @return: The maximum number in the interval [start, end]
 */
func query (root *SegmentTreeNode, start int, end int) int {
    // write your code here
    if root == nil || start > end {
        return 0
    }
    
    if start <= root.Start && root.End <= end {
        return root.Max
    }
    
    return max(query(root.Left, start, end), query(root.Right, start, end))
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
